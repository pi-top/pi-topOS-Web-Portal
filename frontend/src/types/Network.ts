export type Network = {
  ssid: string;
  passwordRequired: boolean
};

export type NetworkCredentials = {
  ssid: string;
  password: string
};
