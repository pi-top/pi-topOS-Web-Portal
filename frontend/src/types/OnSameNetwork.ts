export type OnSameNetworkData = {
  onSameNetwork: boolean;
  piTopIp: string;
  clientIp: string;
}
