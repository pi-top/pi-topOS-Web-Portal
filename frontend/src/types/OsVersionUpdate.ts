export type OsVersionUpdate = {
  shouldBurn: boolean;
  requireBurn: boolean;
  latestOSVersion: string;
  update: boolean;
};
