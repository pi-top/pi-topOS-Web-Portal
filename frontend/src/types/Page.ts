export enum Page {
  Splash = "SPLASH",
  Wifi = "WIFI",
  Upgrade = "UPGRADE",
  Country = "COUNTRY",
  Keyboard = "KEYBOARD",
  Language = "LANGUAGE",
  Registration = "REGISTRATION",
  Restart = "RESTART",
  Terms = "TERMS",
  LandingSplash = "LANDINGSPLASH",
  Links = "LINKS",
}

export enum PageRoute {
  Splash = "/onboarding",
  Wifi = "/onboarding/wifi",
  Upgrade = "/onboarding/upgrade",
  Country = "/onboarding/country",
  Keyboard = "/onboarding/keyboard",
  Language = "/onboarding/language",
  Registration = "/onboarding/registration",
  Restart = "/onboarding/reboot",
  Terms = "/onboarding/terms",
  Links = "/landing/links",
  LandingSplash = "/landing"
}
