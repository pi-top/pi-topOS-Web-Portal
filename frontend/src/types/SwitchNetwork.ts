export type SwitchNetworkData = {
  shouldSwitchNetwork: boolean;
  shouldDisplayDialog: boolean;
  piTopIp: string;
  clientIp: string;
}
