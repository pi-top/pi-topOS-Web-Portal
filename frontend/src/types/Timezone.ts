export type Timezone = {
  countryCode: string;
  timezone: string;
};
