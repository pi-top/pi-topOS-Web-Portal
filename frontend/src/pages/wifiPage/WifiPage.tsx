import React, { useEffect, useState } from "react";

import Layout from "../../components/layout/Layout";
import Select from "../../components/atoms/select/Select";

import wifiPageScreen from "../../assets/images/wifi-screen.png";
import refreshIcon from "../../assets/icons/refresh.svg";
import styles from "./WifiPage.module.css";

import Spinner from "../../components/atoms/spinner/Spinner";
import Button from "../../components/atoms/button/Button";
import ConnectDialogContainer from "./connectDialog/ConnectDialogContainer";
import SkipWarningDialog from "./skipWarningDialog/SkipWarningDialog";
import { Network } from "../../types/Network";
import usePrevious from "../../hooks/usePrevious";

export enum ErrorMessage {
  FetchNetworks = "There was a problem getting networks, please refresh the networks list or skip",
}

export enum ExplanationMessage {
  NotConnected = "Choose from the list below.",
  WiredConnection = "Looks like we're already online. You can still connect to a WiFi network.",
  WiFiConnection = "We've connected successfully.",
}

export type Props = {
  onNextClick?: () => void;
  onSkipClick?: () => void;
  onBackClick?: () => void;
  onRefreshClick: () => void;
  networks: Network[];
  isFetchingNetworks: boolean;
  fetchNetworksError: boolean;
  isConnected: boolean;
  connectedNetwork?: Network;
  setConnectedNetwork: (network: Network) => void;
  skipButtonLabel?: string;
  showSkipWarning?: boolean;
};

export default ({
  onNextClick,
  onSkipClick,
  onBackClick,
  onRefreshClick,
  networks,
  isFetchingNetworks,
  fetchNetworksError,
  isConnected,
  connectedNetwork,
  setConnectedNetwork,
  skipButtonLabel,
  showSkipWarning = true,
}: Props) => {
  const previousConnectedNetwork = usePrevious(connectedNetwork);
  const [selectedNetwork, setSelectedNetwork] = useState(connectedNetwork);
  const [isConnectDialogActive, setIsConnectDialogActive] = useState(false);
  const [isSkipWarningDialogActive, setIsSkipWarningDialogActive] =
    useState(false);

  const { ssid: selectedSSID, bssid: selectedBSSID } = selectedNetwork || {};
  const errorMessage = fetchNetworksError && ErrorMessage.FetchNetworks;
  const getExplanation = () => {
    if (connectedNetwork) {
      return ExplanationMessage.WiFiConnection;
    }
    if (isConnected) {
      return ExplanationMessage.WiredConnection;
    }
    return ExplanationMessage.NotConnected;
  };

  useEffect(() => {
    if (
      connectedNetwork &&
      previousConnectedNetwork?.bssid !== connectedNetwork?.bssid
    ) {
      setSelectedNetwork(connectedNetwork);
    }
  }, [connectedNetwork, previousConnectedNetwork]);

  return (
    <>
      <Layout
        banner={{
          src: wifiPageScreen,
          alt: "wifi-page-banner",
        }}
        prompt={
          <>
            Connect me to <span className="green">WiFi</span>
          </>
        }
        explanation={getExplanation()}
        nextButton={{
          onClick: onNextClick,
          disabled: !isConnected,
        }}
        skipButton={{
          label: skipButtonLabel,
          onClick: () => {
            if (!isConnected && showSkipWarning) {
              setIsSkipWarningDialogActive(true);
            } else if (onSkipClick) {
              onSkipClick();
            }
          },
        }}
        backButton={{ onClick: onBackClick }}
        showBack={!!onBackClick}
        showNext={!!onNextClick}
        showSkip={!!onSkipClick}
      >
        <div className={styles.wifiSelectContainer}>
          <Select
            // force rerender when selected ssid changes
            key={selectedBSSID}
            value={
              selectedBSSID &&
              selectedSSID && {
                value: selectedBSSID,
                label: selectedSSID,
              }
            }
            options={networks.map(({ ssid, bssid }) => ({
              value: bssid,
              label: ssid,
            }))}
            onChange={(newBSSID) => {
              setIsConnectDialogActive(true);
              setSelectedNetwork(
                networks.find((network) => newBSSID === network.bssid)
              );
            }}
            placeholder="Please select WiFi network..."
            isDisabled={isFetchingNetworks}
          />

          <div className={styles.selectDetailsContainer}>
            {isFetchingNetworks ? (
              <>
                <Spinner size={20} />{" "}
                <span className={styles.settingMessage}>
                  fetching networks...
                </span>
              </>
            ) : (
              <Button
                unstyled
                onClick={onRefreshClick}
                className={styles.refreshButton}
              >
                <img src={refreshIcon} alt="refresh-button" />
              </Button>
            )}
          </div>
        </div>

        {errorMessage && <span className={styles.error}>{errorMessage}</span>}
      </Layout>

      <ConnectDialogContainer
        active={isConnectDialogActive}
        network={selectedNetwork}
        setConnectedNetwork={setConnectedNetwork}
        onCancel={() => {
          setIsConnectDialogActive(false);
          setSelectedNetwork(undefined);
        }}
        onDone={() => {
          setIsConnectDialogActive(false);
          if (isConnected && onNextClick) {
            onNextClick();
          }
        }}
      />
      <SkipWarningDialog
        active={isSkipWarningDialogActive}
        onConnectClick={() => setIsSkipWarningDialogActive(false)}
        onSkipClick={onSkipClick || (() => {})}
      />
    </>
  );
};
