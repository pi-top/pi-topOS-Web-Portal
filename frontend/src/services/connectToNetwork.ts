import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

import { NetworkCredentials } from "../types/Network";


export default async function connectToNetwork(networkCredentials: NetworkCredentials, timeout?: number) {
  await axios.post(
    `${apiBaseUrl}/wifi-credentials`,
    networkCredentials,
    { timeout: timeout || 10000 }
  );
}
