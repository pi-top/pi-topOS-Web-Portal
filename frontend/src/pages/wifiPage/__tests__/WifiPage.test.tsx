import React from "react";
import {
  render,
  fireEvent,
  BoundFunction,
  GetByBoundAttribute,
  QueryByText,
  GetByText,
  RenderResult,
  QueryByBoundAttribute,
} from "@testing-library/react";
import { AllByText, wait } from "@testing-library/dom";

import WifiPage, { Props, ErrorMessage, ExplanationMessage } from "../WifiPage";
import queryReactSelect from "../../../../test/helpers/queryReactSelect";
import reactSelectIsDisabled from "../../../../test/helpers/reactSelectIsDisabled";
import { KeyCode } from "../../../../test/types/Keys";
import { Network } from '../../../types/Network';
import querySpinner from "../../../../test/helpers/querySpinner";
import connectToNetwork from "../../../services/connectToNetwork";

jest.mock("../../../services/connectToNetwork");
const connectToNetworkMock = connectToNetwork as jest.Mock;

describe("WifiPage", () => {
  let defaultProps: Props;
  let wifiPage: HTMLElement;
  let getByAltText: BoundFunction<GetByBoundAttribute>;
  let queryByText: BoundFunction<QueryByText>;
  let getByText: BoundFunction<GetByText>;
  let queryAllByText: BoundFunction<AllByText>;
  let queryByTestId: BoundFunction<QueryByBoundAttribute>;
  let queryByAltText: BoundFunction<QueryByBoundAttribute>;
  let queryByLabelText: BoundFunction<QueryByBoundAttribute>;
  let getByLabelText: BoundFunction<GetByBoundAttribute>;
  let rerender: RenderResult["rerender"];
  beforeEach(() => {
    defaultProps = {
      networks: [],
      onNextClick: jest.fn(),
      onSkipClick: jest.fn(),
      onBackClick: jest.fn(),
      onRefreshClick: jest.fn(),
      setConnectedNetwork: jest.fn(),
      isFetchingNetworks: false,
      fetchNetworksError: false,
      isConnected: false,
      isCompleted: false,
    };

    connectToNetworkMock.mockResolvedValue('OK')

    ({
      container: wifiPage,
      getByAltText,
      queryByText,
      queryAllByText,
      queryByTestId,
      queryByAltText,
      queryByLabelText,
      getByLabelText,
      getByText,
      rerender,
    } = render(<WifiPage {...defaultProps} />));
  });
  afterEach(() => {
    connectToNetworkMock.mockRestore();
  })

  it("renders the correct banner image", () => {
    expect(queryByAltText("wifi-page-banner")).toMatchSnapshot();
  });

  it("renders prompt correctly", () => {
    const prompt = wifiPage.querySelector(".prompt");
    expect(prompt).toMatchSnapshot();
  });

  it("Next button is disabled on no internet", () => {
    expect(getByText("Next")).toHaveProperty("disabled");
  });

  it("calls onBackClick when back button clicked", () => {
    fireEvent.click(getByText("Back"));

    expect(defaultProps.onBackClick).toHaveBeenCalled();
  });

  it("calls onRefreshClick when refresh button clicked", () => {
    fireEvent.click(getByAltText("refresh-button"));

    expect(defaultProps.onRefreshClick).toHaveBeenCalled();
  });

  it("renders select", () => {
    expect(queryReactSelect(wifiPage)).toBeInTheDocument();
  });

  it("renders correct placeholder in select", () => {
    expect(getByText("Please select WiFi network...")).toBeInTheDocument();
  });

  it("renders refresh networks button", () => {
    expect(queryByAltText("refresh-button")).toBeInTheDocument();
  });

  it("doesn't render the fetching networks message", () => {
    expect(queryByText("fetching networks...")).not.toBeInTheDocument();
  });

  it("doesn't render the fetching networks spinner", () => {
    expect(querySpinner(wifiPage)).not.toBeInTheDocument();
  });

  it("renders the dialog", () => {
    expect(queryByTestId("dialog")).toBeInTheDocument();
  });

  it("hides the connect dialog", () => {
    expect(queryByTestId("dialog")).toHaveClass("hidden");
  });

  describe("when entering page with an internet connection", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        isConnected: true,
      };

      rerender(<WifiPage {...defaultProps} />);
    });

    it("renders a text saying that the user is already connected", () => {
      expect(
        queryByText(ExplanationMessage.WiredConnection)
      ).toBeInTheDocument();
    });

    it("calls onNextClick when next button clicked", () => {
      fireEvent.click(getByText("Next"));

      expect(defaultProps.onNextClick).toHaveBeenCalled();
    });

    it("skip button exists", () => {
      expect(queryByText("Skip")).toBeInTheDocument();
    });

    it("calls onSkipClick when skip button clicked", () => {
      fireEvent.click(getByText("Skip"));

      expect(defaultProps.onSkipClick).toHaveBeenCalled();
    });
  });

  describe("when entering page without an internet connection", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        isConnected: false,
      };

      rerender(<WifiPage {...defaultProps} />);
    });

    it("renders the explanation", () => {
      expect(queryByText(ExplanationMessage.NotConnected)).toBeInTheDocument();
    });
  });

  describe("when is fetching networks", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        isFetchingNetworks: true,
      };

      rerender(<WifiPage {...defaultProps} />);
    });

    it("renders the fetching networks message", () => {
      expect(queryByText("fetching networks...")).toBeInTheDocument();
    });

    it("renders the fetching networks spinner", () => {
      expect(querySpinner(wifiPage)).toBeInTheDocument();
    });

    it("the networks select is disabled", () => {
      expect(reactSelectIsDisabled(wifiPage)).toEqual(true);
    });

    it("does not render refresh networks button", () => {
      expect(queryByAltText("refresh-button")).not.toBeInTheDocument();
    });
  });

  describe("when fetchNetworkError is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        fetchNetworksError: true,
      };

      rerender(<WifiPage {...defaultProps} />);
    });

    it("renders error message", () => {
      expect(queryByText(ErrorMessage.FetchNetworks)).toBeInTheDocument();
    });

    it("skip button is present", () => {
      expect(getByText("Skip")).toBeInTheDocument();
    });

    it("calls onSkipClick when skip button clicked", () => {
      fireEvent.click(getByText("Skip"));

      expect(defaultProps.onSkipClick).toHaveBeenCalled();
    });
  });

  describe("when networks are passed", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        networks: [
          {
            ssid: "password-protected-ssid",
            passwordRequired: true,
          },
          {
            ssid: "unsecured-ssid",
            passwordRequired: false,
          },
        ],
      };

      rerender(<WifiPage {...defaultProps} />);
    });

    it("renders correct select options", () => {
      fireEvent.keyDown(queryReactSelect(wifiPage)!, {
        keyCode: KeyCode.DownArrow,
      });

      defaultProps.networks.forEach(({ ssid }) => {
        expect(queryByText(ssid)).toBeInTheDocument();
      });
    });

    it("allows selecting option", () => {
      const {
        networks: [firstNetwork, ...remainingNetworks],
      } = defaultProps;

      // open the select
      fireEvent.keyDown(queryReactSelect(wifiPage)!, {
        keyCode: KeyCode.DownArrow,
      });

      // click the wifi option
      fireEvent.click(getByText(firstNetwork.ssid));

      /*
        the menu closes and the selected option should still be rendered
        but the second option should not be rendered as the menu is closed
      */
      expect(queryAllByText(firstNetwork.ssid).length).toBeGreaterThan(0);
      remainingNetworks.forEach(({ ssid }) => {
        expect(queryByText(ssid)).not.toBeInTheDocument();
      });
    });

    describe("when option without a password is selected", () => {
      let network: Network;
      beforeEach(() => {
        // open the select
        fireEvent.keyDown(queryReactSelect(wifiPage)!, {
          keyCode: KeyCode.DownArrow,
        });

        network = defaultProps.networks.find(
          ({ passwordRequired }) => !passwordRequired
        )!;

        // click the wifi option
        fireEvent.click(getByText(network.ssid));
      });

      it("shows the connect dialog", () => {
        expect(queryByTestId("dialog")).not.toHaveClass("hidden");
      });

      it("renders dialog message correctly", () => {
        expect(queryByTestId("dialog-message")).toMatchSnapshot();
      });

      describe("on cancel click", () => {
        beforeEach(() => {
          fireEvent.click(getByText("Cancel"));
        });

        it("hides the dialog", () => {
          expect(queryByTestId("dialog")).toHaveClass("hidden");
        });

        it("resets selected network", () => {
          expect(queryByText(network.ssid)).not.toBeInTheDocument();
        });
      });

      describe('when finished connecting to network', () => {
        beforeEach(async () => {
          fireEvent.click(getByText('Join'));

          await wait();
        });

        it('hides the connect dialog on OK click', () => {
          fireEvent.click(getByText('OK'));

          expect(queryByTestId("dialog")).toHaveClass("hidden");
        })
      })
    });
  });
});
