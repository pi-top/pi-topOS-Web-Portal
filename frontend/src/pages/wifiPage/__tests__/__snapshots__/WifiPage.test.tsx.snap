// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WifiPage renders prompt correctly 1`] = `
<h1
  class="prompt"
>
  Connect me to 
  <span
    class="green"
  >
    WiFi
  </span>
</h1>
`;

exports[`WifiPage renders the correct banner image 1`] = `
<img
  alt="wifi-page-banner"
  class="image"
  draggable="false"
  src="wifi-screen.png"
  style="max-height: 100%; max-width: 100%;"
/>
`;

exports[`WifiPage when networks are passed when option without a password is selected renders dialog message correctly 1`] = `
<h3
  class="messageContainer"
  data-testid="dialog-message"
>
  The WiFi network 
  <span
    class="green"
  >
    unsecured-ssid
  </span>
   
  does not require
   a password
</h3>
`;
